reframe/core/schedulers/firecrest.py

```diff
@@ -0,0 +1,364 @@
+                                           slurm_state_completed)
+from reframe.core.exceptions import JobSchedulerError
+
+if sys.version_info >= (3, 7):
+    import firecrest as fc
+
+
+def join_and_normalize(*args):
+    joined_path = os.path.join(*args)
+    normalized_path = os.path.normpath(joined_path)
+    return normalized_path
+
+
+class _SlurmFirecrestJob(sched.Job):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._is_array = False
+        self._is_cancelling = False
+        self._remotedir = None
+        self._localdir = None
+
+        # The compacted nodelist as reported by Slurm. This must be updated
+        # in every poll as Slurm may be slow in reporting the exact nodelist
+        self._nodespec = None
+        self._stage_prefix = rt.runtime().stage_prefix
+
+    @property
+    def is_array(self):
+        return self._is_array
+
+    @property
+    def is_cancelling(self):
+        return self._is_cancelling
+
+
+@register_scheduler('firecrest-slurm')
+class SlurmFirecrestJobScheduler(SlurmJobScheduler):
+    def __init__(self, *args, **kwargs):
+        def set_mandatory_var(var):
+            res = os.environ.get(var)
+            if res:
+                return res
+
+            raise JobSchedulerError(f'the env var {var} is mandatory for the '
+                                    f'firecrest scheduler')
+
+        if sys.version_info < (3, 7):
+            raise JobSchedulerError('the firecrest scheduler needs '
+                                    'python>=3.7')
+
+        super().__init__(*args, **kwargs)
+        client_id = set_mandatory_var("FIRECREST_CLIENT_ID")
+        client_secret = set_mandatory_var("FIRECREST_CLIENT_SECRET")
+        token_uri = set_mandatory_var("AUTH_TOKEN_URL")
+        firecrest_url = set_mandatory_var("FIRECREST_URL")
+        self._system_name = set_mandatory_var("FIRECREST_SYSTEM")
+        self._remotedir_prefix = set_mandatory_var('FIRECREST_BASEDIR')
+
+        # Setup the client for the specific account
+        self.client = fc.Firecrest(
+            firecrest_url=firecrest_url,
+            authorization=fc.ClientCredentialsAuth(client_id, client_secret,
+                                                   token_uri)
+        )
+
+        params = self.client.parameters()
+        for p in params['utilities']:
+            if p['name'] == 'UTILITIES_MAX_FILE_SIZE':
+                self._max_file_size_utilities = float(p['value'])*1000000
+                break
+
+        self._local_filetimestamps = {}
+        self._remote_filetimestamps = {}
+        self._cleaned_remotedirs = set()
+
+    def make_job(self, *args, **kwargs):
+        return _SlurmFirecrestJob(*args, **kwargs)
+
+    def _push_artefacts(self, job):
+        def _upload(local_path, remote_path):
+            f_size = os.path.getsize(local_path)
+            if f_size <= self._max_file_size_utilities:
+                self.client.simple_upload(
```

Comment:
Does it make sense to add a log entry here stating that this is synchronous upload?